### DISCUSSION TITLE
Reflect forecasted data in notification in-between update intervals

### DISCUSSION THREAD

**Discussion by jayb-g** [breezy-weather/breezy-weather#1272] [Ideas]
### Describe your idea

Currently the app updates the main screen and the forecast notification only when the background refresh interval is hit, which is by default 1.5 hours. While doing this app doesn't fully utilize the already fetched hourly forecast data to show live weather(even if its forecasted, and not really live) in-between the refresh interval.

Let's say forecast for today is
9.00 - Partly cloudy
10.00 - Rain
11.00 - Overcast
12.00 - Overcast
13.00 - Thunderstorm - Heavy Rain

Now if the app had updated data at say 8.50, next update would be at 10.20. Till then it will keep showing Partly cloudy and not Rain after 10.00
Similarly, it will keep showing overcast until 13.20.

It should instead show the weather forecast closest to the current time. Maybe the app can have an internal refresh rate of 15 mins for that?

Note: I understand there's a refresh rate setting for foreast update. But I'm not talking about that. Default of 1.5hours seems just fine. I'm suggesting an impvoement in how the already fetched forecast data can be better used, even with longer refresh intervals.

### Other details

@papjul has already replied on the issue #1269 :

From my understanding, you would like to use observations data for current for -let’s say- 15 min, then after that period, fallback to hourly forecast to not use past observations.

At first sight, I see a few difficulties:

- Hourly forecast is not that reliable. You can see that just by freshly refreshing and comparing your hourly forecast with the current observations. There is sometimes a huge gap because forecast and observations by stations don't work the same way at all. (For sources with no observations, such as Open-Meteo, however, it will be the same data)
- For some sources forecast for 9:00 can mean "forecast from 8:00 to 9:00", for others "forecast from 8:30 to 9:30" and for others "forecast from 9:00 to 10:00", so it's hard to know which one to pick in the situation you describe
- The refresh of notifications and widgets happen only when there is an actual refresh of weather data, this would mean adding an additional process -let’s say- every 15 min, which Android doesn't like much when there are too many works scheduled and with a frequent rate. There is a high risk of being killed on some devices such as Xiaomi, Samsung or others

If you would like to discuss it further, please fill in the Idea template: https://github.com/breezy-weather/breezy-weather/discussions/new?category=ideas

### Acknowledgements

- [X] I have searched the [existing issues](https://github.com/breezy-weather/breezy-weather/issues) and [existing discussions](https://github.com/breezy-weather/breezy-weather/discussions/categories/ideas) and this is a new ticket, **NOT** a duplicate or related to another open OR closed issue or discussion.
- [X] The idea I’m suggesting is not in the [list of features that will not be implemented](https://github.com/breezy-weather/breezy-weather/blob/main/README.md#features-that-will-not-be-implemented).
- [X] I have updated the app to **[version 5.2.7 or later](https://github.com/breezy-weather/breezy-weather/releases/latest)**. Even if I’m getting updates from F-Droid default repo, I don’t need to uninstall/reinstall to update.
- [x] I wrote my request in English, and any screenshot I attached are also in English (language can be changed from Settings > Appearance). I can write in my native language under the English text, if necessary.
- [X] I understand that this app is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY, so I may not get a positive reply to my request, if any at all.

**Comment by jayb-g** [comment]
> Hourly forecast is not that reliable. You can see that just by freshly refreshing and comparing your hourly forecast with the current observations. 

I agree. But showing hourly forecasted data as current weather is better than showing the previously fetched an hour or older data as current weather.

> There is sometimes a huge gap because forecast and observations by stations don't work the same way at all

Wouldn't it still be better in that case to show forecasted data instead of waiting for another hour for it to fetch and show live data?

**Comment by jayb-g** [comment]
> The refresh of notifications and widgets happen only when there is an actual refresh of weather data, this would mean adding an additional process -let’s say- every 15 min, which Android doesn't like much when there are too many works scheduled and with a frequent rate. There is a high risk of being killed on some devices such as Xiaomi, Samsung or others

Would keeping it longer than 15mins, maybe 30 mins be ok? For notification and widget update? Basically anything less than the default refresh interval(1.5hours) is better. Ideally 45mins at max.(90/2)

**Reply by papjul** [reply]
The hourly forecast -which you want to use- is a forecast valid for a full 1 hour.
In that case, my recommendation would be to lower your refresh rate from 1.5 hours to 1 hour.
It's a good compromise to get an accurate observation every hour, rather than a -not very logic- mix of observation and forecast every hour (such as what I described in another comment), which would on top of that, add a non-negligible layer of complexity to the app.
The default refresh rate will remain 1.5 hours though, as most users don't have a need to check the outside temperature so often (or they can just manually refresh if they need it once or twice a day).

**Comment by jayb-g** [comment]
Without such a feature, I more often see the outdated(data of previous hour) weather data in the notification until I open the app and let it refresh or do it manually. So if its 4pm, the app is showing weather data of 3pm or even 2pm depending on when it was last refreshed.

**Reply by papjul** [reply]
So, let's say you refresh every 1.5 hours, is this what you would like?
- 09:00 - observation from station (refresh in the background)
- 09:30 - forecast from 09:00? Would it be really more accurate than the observation from 09:00?
- 10:00 - forecast from 10:00?
- 10:30 - observation from station (refresh in the background)
- 11:00 - forecast from 11:00
- 11:30 - forecast from 11:00 again?
- 12:00 - observation from station (refresh in the background)
- 12:30 - forecast from 12:00? Would it be really more accurate than the observation from 12:00?
- and so on…

**Reply by papjul** [reply]
Now, let's say you follow my recommendation to lower to a 1 hour refresh rate, this is what would happen:
- 09:00 - observation from station (refresh in the background)
- 09:30 - forecast from 09:00? Would it be really more accurate than the observation from 09:00?
- 10:00 - observation from station (refresh in the background)
- 10:30 - forecast from 10:00? Would it be really more accurate than the observation from 10:00?
- 11:00 - observation from station (refresh in the background)
- 11:30 - forecast from 11:00? Would it be really more accurate than the observation from 11:00?
- 12:00 - observation from station (refresh in the background)
- 12:30 - forecast from 12:00? Would it be really more accurate than the observation from 12:00?
- and so on…

Now it makes much more sense to just keep the observation all the time.

**Reply by jayb-g** [reply]
yes i mentioned default of 1.5hours is just fine. changing it to 1 hour is also not a problem for me(which would actually be better as shown below)
Observations are also updated at set refresh intervals right? I'm using open-meteo so observations are not a factor but still for sake of general discussion, let's consider it.

Does the app have fixed slots for background refresh ie at full or half hours? ie 9.00 and then 10.30

I don't know. But I think it does not. Here its assumed that it refreshes at 9.00 but in reality it may have refreshed at 9.10 which will mean that at 1.5 hours refresh interval, it will still show weather data(and observation from) of 9'o clock until 10.40.
So on so forth.

@10.30 - forecast for 10.00(as updated at 9.10 along with observation)? Would it be really more accurate than the observation from 09:00? 
Definitely or ideally mostly yes.

Although if I just change the background refresh interval to 1 hour. The concerns go away. But I may not want it to cost 1.5 times the resources. 

Anyway I understand if for whatever reason its not feasible to implement. Just wanted to share the idea if it can be of any help.

Thanks to all contributors for this awesome app.

**Reply by papjul** [reply]
> I'm using open-meteo so observations are not a factor

If I recall correctly, Open-Meteo uses "15-minutely" for current, so it should still be better than hourly forecast.

> Does the app have fixed slots for background refresh ie at full or half hours? ie 9.00 and then 10.30

No, Android is in charge of launching the work. We ask a 1.5 hour rate, but they don't guarantee it. It can be sooner or later to ensure a smooth experience on your device (i.e. avoid running all works scheduled by all apps at the same time while you're playing a video game).

> Although if I just change the background refresh interval to 1 hour. The concerns go away. But I may not want it to cost 1.5 times the resources.

From user side, it's a very lightweight process, both in network consumption and in processing it. Loading an API URL is just a few Ko, while just opening a web page on a newspaper website can easily be a few Mo, so multiply this with each page, and browsing Internet is way more heavy.

From server side, yes, it will cost an additional call, but if you need it, you need it. Just make sure to turn off Internet on your phone while you sleep to avoid useless network consumption (from Breezy Weather or any other app). Android will run all works depending on network as soon as network is up again.

**Comment by jayb-g** [comment]
Also, I believe updating widget data without using the network shouldn't be a problem for battery at all. Many stock widgets have 15mins interval by default and they don't have any battery usage issues and they even use network at those intervals. Maybe they don't run when phone is in sleep mode?

**Reply by papjul** [reply]
It doesn't matter the task is not battery-heavy. Many vendors don't like apps scheduling works.
Stock widgets are probably excluded from such restrictions.
